packages/compass-crud/src/components/insert-document-dialog.tsx

```diff
@@ -63,117 +70,180 @@ export type InsertDocumentDialogProps = InsertCSFLEWarningBannerProps & {
+    return invalidElements.length > 0;
+  }, [invalidElements, jsonDoc, jsonView]);
+
+  const handleInvalid = useCallback(
+    (el: Element) => {
+      if (!invalidElements.includes(el.uuid)) {
+        setInvalidElements((elements) => [...elements, el.uuid]);
+      }
+    },
+    [invalidElements]
+  );
+
+  const handleValid = useCallback(
+    (el: Element) => {
+      if (hasErrors()) {
+        setInvalidElements((invalidElements) =>
+          without(invalidElements, el.uuid)
         );
+      }
+    },
+    [hasErrors, setInvalidElements]
+  );
+
+  useEffect(() => {
+    if (isOpen && track) {
+      track('Screen', { name: 'insert_document_modal' }, undefined);
     }
+  }, [isOpen, track]);
 
-    if (this.props.isOpen && !this.hasManyDocuments()) {
-      if (prevProps.jsonView && !this.props.jsonView) {
+  const prevJsonView = useRef(jsonView);
+  useEffect(() => {
+    const viewHasChanged = prevJsonView.current !== jsonView;
+    prevJsonView.current = jsonView;
+    if (isOpen && !hasManyDocuments() && viewHasChanged) {
+      if (!jsonView) {
         // When switching to Hadron Document View.
         // Reset the invalid elements list, which contains the
         // uuids of each element that has BSON type cast errors.
-        this.invalidElements = [];
+        setInvalidElements([]);
         // Subscribe to the validation errors for BSON types on the document.
-        this.props.doc.on(Element.Events.Invalid, this.handleInvalid);
-        this.props.doc.on(Element.Events.Valid, this.handleValid);
-      } else if (!prevProps.jsonView && this.props.jsonView) {
+        doc.on(Element.Events.Invalid, handleInvalid);
+        doc.on(Element.Events.Valid, handleValid);
+      } else {
         // When switching to JSON View.
         // Remove the listeners to the BSON type validation errors in order to clean up properly.
-        this.props.doc.removeListener(
-          Element.Events.Invalid,
-          this.handleInvalid
-        );
-        this.props.doc.removeListener(Element.Events.Valid, this.handleValid);
+        doc.removeListener(Element.Events.Invalid, handleInvalid);
+        doc.removeListener(Element.Events.Valid, handleValid);
       }
     }
+  }, [isOpen, jsonView, doc, handleValid, handleInvalid, hasManyDocuments]);
 
-    if (state.insertInProgress) {
-      this.setState({ insertInProgress: false });
+  useEffect(() => {
+    if (insertInProgress) {
+      setInsertInProgress(false);
     }
-  }
+  }, [insertInProgress]);
 
-  componentWillUnount() {
-    if (!this.hasManyDocuments()) {
+  useEffect(() => {
+    if (!hasManyDocuments() && doc) {
       // When closing the modal.
       // Remove the listeners to the BSON type validation errors in order to clean up properly.
-      this.props.doc.removeListener(Element.Events.Invalid, this.handleInvalid);
-      this.props.doc.removeListener(Element.Events.Valid, this.handleValid);
+      doc.removeListener(Element.Events.Invalid, handleInvalid);
```

Comment:
Should these be removing the listeners in a returned function since we want this on unmount?
Also should it be using `isOpen` to manage the cleanup then as well since we're now keeping this modal around when `isOpen` is set to false?